### PR TITLE
Feature/cargo push

### PR DESCRIPTION
**Pull request by jsvisa**
@ddysher @ijumps PTAL

### CODE REVIEW

**Comment by deyuanix** [comment]
@ijump

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM ubuntu:16.04
+MAINTAINER Delweng Zheng <delweng@gmail.com>
+
+RUN apt-get update && apt-get install -y cron
```
Comment (ijumps) [inline Dockerfile]: multi `RUN` to one `RUN` to reduce docker image layer, e.g.

```
##update system
RUN apt-get update && apt-get upgrade -y \

##runtime install
	&& apt-get install -y nginx \

##clean
	&& apt-get autoremove -y \
	&& rm -rf /var/lib/apt/lists/*
```

File: heketi-start.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
+#!/bin/sh
```
Comment (ijumps) [inline heketi-start.sh:1]: If you use a bash script as an entrypoint, a good practice is to write script like this:

```
#!/bin/bash
set -e

#some setup here

exec "$@"
```

A shell script will not handle linux signal well, and this shell will become pid 1. call `exec` will let the real entrypoint as pid 1

File: heketi-start.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    case $1 in
+        -c|--config)
+            config=$2; shift 2;;
+        -b|--backup)
+            backup=$2; shift 2;;
+        -d|--datadb)
+            datadb=$2; shift 2;;
+        --)
+            shift; break;;
+        -h|--help)
```
Comment (ijumps) [inline heketi-start.sh:24]: `-h` not work

File: heketi-start.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+config=${config-/etc/heketi/heketi.json}
+backup=${backup-/backupdb/}
+datadb=${datadb-/var/lib/heketi/heketi.db}
+datadb_dir=$(dirname $datadb)
+
+mkdir -p $backup
+if [ ! -f $datadb ]; then
+    recent=$(ls $backup | sort -r | head -1)
+    if [ $recent -ne "" ]; then
```
Comment (ijumps) [inline heketi-start.sh]: `$recent -ne ""` to `"$recent" != ""`

File: Makefile [hunk 5]
```diff
@@ -90,6 +90,15 @@ clean:
 	rm -rf dist
 	@$(MAKE) -C client/cli/go clean
 
+docker:
+ifdef DOCKER_REGISTRY
+	@echo "Push image $(APP_NAME):$(VERSION) to $(DOCKER_REGISTRY)/caicloud/$(APP_NAME):$(VERSION:v%=%)"
+	docker build -t $(DOCKER_REGISTRY)/caicloud/$(APP_NAME):$(VERSION:v%=%) .
+else
+	@echo "Push image $(APP_NAME):$(VERSION) to default docker registry(default is hub.docker.com)"
+	docker build -t caicloud/$(APP_NAME):$(VERSION:v%=%) .
```
Comment (ijumps) [inline Makefile]: we usually keep image prefix `v`, e.g. `v1.2.3`

File: Makefile [hunk 6]
```diff
@@ -90,6 +90,15 @@ clean:
 	rm -rf dist
 	@$(MAKE) -C client/cli/go clean
 
+docker:
+ifdef DOCKER_REGISTRY
+	@echo "Push image $(APP_NAME):$(VERSION) to $(DOCKER_REGISTRY)/caicloud/$(APP_NAME):$(VERSION:v%=%)"
```
Comment (ijumps) [inline Makefile]: this only do docker build but not docker push

File: Dockerfile (hunk 1 above)
Comment (jsvisa) [reply Dockerfile]: @ijumps fixed

File: heketi-start.sh (hunk 2 above)
Comment (jsvisa) [reply heketi-start.sh:1]: @ijumps  thanks for advise

File: heketi-start.sh (hunk 3 above)
Comment (jsvisa) [reply heketi-start.sh:24]: @ijumps fixed

File: heketi-start.sh (hunk 4 above)
Comment (jsvisa) [reply heketi-start.sh]: @ijumps thanks for advise

File: Makefile (hunk 6 above)
Comment (jsvisa) [reply Makefile]: @ijumps added

File: Makefile (hunk 5 above)
Comment (jsvisa) [reply Makefile]: fixed @ijumps

**Comment by deyuanix** [comment]
@ijumps can we merge can proceed?

File: heketi-start.sh [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+backup=${backup-/backupdb/}
+datadb=${datadb-/var/lib/heketi/heketi.db}
+datadb_dir=$(dirname $datadb)
+
+mkdir -p $backup
+if [ ! -f $datadb ]; then
+    recent=$(ls $backup | sort -r | head -1)
+    if [ "$recent" != "" ]; then
+        tar zxf $backup/$recent -C $datadb_dir
+        ret=$?
```
Comment (ijumps) [inline heketi-start.sh:39]: any error will cause script exit when `set -e`, for this script, `ret=$?` will never reach if `tar zxf $backup/$recent -C $datadb_dir` failed. So `set -e` may not need here.

File: heketi-start.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        echo "Recovery database from $backup/$recent to $datadb"
+    else
+        echo "Heketi fresh starting!"
+    fi
+fi
+
+echo "30 2 * * * root (cd $datadb_dir; export backup_file=heketi-db-\$(date +\\%F-\\%H-\\%M).tar.gz && tar zcf \$backup_file heketi.db && mv \$backup_file $backup) >/dev/null 2>&1\n#" > /etc/crontab
+
+pidof cron
+if [ $? != 0 ]; then
```
Comment (ijumps) [inline heketi-start.sh:53]: the same

File: heketi-start.sh (hunk 7 above)
Comment (jsvisa) [reply heketi-start.sh:39]: @ijumps  fixed  https://github.com/caicloud/heketi/commit/ff0551823f03bced72da13892d48b3e5ea9f97dd

**Comment by ijumps** [comment]
LGTM

**Pull request merged**